dynamiqs/time_tensor.py

```diff
@@ -57,14 +59,55 @@ def totime(
                 f' a square matrix, but has shape {tuple(tensor.shape)}.'
             )
 
-        factors = [_PWCTensor(times, values)]
+        factors = [_PWCFactor(times, values)]
         tensors = tensor.unsqueeze(0)  # (1, n, n)
         return PWCTimeTensor(factors, tensors)
-    # constant time tensor
+    # === modulated time tensor
+    if isinstance(x, tuple) and len(x) == 2:
+        factor, tensor = x
+
+        # check factor
+        if not callable(factor):
+            raise TypeError(
+                'For a modulated time tensor `(factor, tensor)`, argument `factor` must'
+                f' be a callable, but has type {obj_type_str(factor)}.'
+            )
+        factor0 = factor(0.0)
+        if not isinstance(factor0, Tensor):
+            raise TypeError(
+                'For a modulated time tensor `(factor, tensor)`, argument `factor` must'
+                f' return a tensor, but returns type {obj_type_str(factor0)}.'
+            )
+        if factor0.device != device:
+            raise TypeError(
+                'For a modulated time tensor, the tensor returned by the callable must'
+                f' be on device `{device}, but is on device `{factor0.device}`. This is'
+                ' to avoid costly device transfer at each solver time step.'
```

Comment:
```suggestion
                ' necessary to avoid costly device transfer at each solver time step.'
```
And in the other comment too.